pkg/controllermanager/controller/plant/plant_health.go

```diff
@@ -0,0 +1,137 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plant
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"k8s.io/client-go/discovery"
+
+	"github.com/gardener/gardener/pkg/utils/kubernetes/health"
+
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+
+	gardencorev1alpha1 "github.com/gardener/gardener/pkg/apis/core/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/gardener/gardener/pkg/apis/core/v1alpha1/helper"
+)
+
+// NewHealthCheker creates a new health checker.
+func NewHealthCheker(plantClient client.Client, discoveryClient discovery.DiscoveryInterface) *HealthChecker {
+	return &HealthChecker{
+		plantClient:     plantClient,
+		discoveryClient: discoveryClient,
+	}
+}
+
+// HealthChecker contains the condition thresholds.
+type HealthChecker struct {
+	plantClient     client.Client
+	discoveryClient discovery.DiscoveryInterface
+}
+
+// CheckPlantClusterNodes checks whether cluster nodes in the given listers are complete and healthy.
+func (h *HealthChecker) CheckPlantClusterNodes(ctx context.Context, condition *gardencorev1alpha1.Condition) (*gardencorev1alpha1.Condition, error) {
+	nodeList := &corev1.NodeList{}
+	err := h.plantClient.List(ctx, &client.ListOptions{}, nodeList)
+	if err != nil {
+		return nil, err
+	}
+
+	if exitCondition := h.checkNodes(*condition, nodeList); exitCondition != nil {
+		return exitCondition, nil
+	}
+
+	updatedCondition := helper.UpdatedCondition(*condition, gardencorev1alpha1.ConditionTrue, string(gardencorev1alpha1.PlantEveryNodeReady), "Every node registered to the cluster is ready")
+	return &updatedCondition, nil
+}
+
+// CheckAPIServerAvailability checks if the API server of a Plant cluster is reachable and measure the response time.
+func (h *HealthChecker) CheckAPIServerAvailability(condition gardencorev1alpha1.Condition) gardencorev1alpha1.Condition {
+	// Try to reach the Plant API server and measure the response time.
+	now := time.Now()
+	discoveryClient, ok := h.discoveryClient.(discovery.DiscoveryInterface)
+	if !ok {
+		message := fmt.Sprintf("discoveryClient does not implement discovery interface")
+		return helper.UpdatedCondition(condition, gardencorev1alpha1.ConditionFalse, "HealthzRequestFailed", message)
+	}
+	response := discoveryClient.RESTClient().Get().AbsPath("/healthz").Do()
```

Comment:
This is just copied from our current health checking process - please re-use or factor out the components.